### PR TITLE
feat(sql): make trailing comma optional

### PR DESCRIPTION
**Pull request by bluestreak01**
Fixes https://github.com/questdb/questdb/issues/3676
Closes https://github.com/questdb/questdb/issues/3666

Relax SQL syntax when it comes to trailing commas such as:

```sql
select
    a,
    b,
    c,
    d,
from tab
```

Such commas are ignored and not reported as errors, allowing hassle-free reodering of selected columns in editor.

At the same added stricter syntax around aliasing wildcarded columns, such as:

```sql
select a.* b from tab
```

Alias `b` makes no sense and will be reported as syntax error.

Also fixed `master` regression https://github.com/questdb/questdb/issues/3676

### CODE REVIEW

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 33 / 34 (97.06%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/SqlOptimiser.java|3|4|75.00%|
|:large_blue_circle:|io/questdb/griffin/SqlParser.java|27|27|100.00%|
|:large_blue_circle:|io/questdb/griffin/WhereClauseParser.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/model/ExpressionNode.java|2|2|100.00%|

File: core/src/main/java/io/questdb/griffin/SqlOptimiser.java [hunk 1]
```diff
@@ -3437,6 +3437,9 @@ private void rewriteNegativeLimit(QueryModel model, SqlExecutionContext executio
                 int timestampKey = current.getColumnNameToAliasMap().keyIndex(nested.getTimestamp().token);
                 if (timestampKey < 0) {
                     int timestampIdx = current.getColumnAliasIndex(current.getColumnNameToAliasMap().valueAt(timestampKey));
+                    if (timestampIdx == -1) {
```
Comment (marregui) [inline core/src/main/java/io/questdb/griffin/SqlOptimiser.java:3440]: whoooooops left in there by mistake?

File: core/src/main/java/io/questdb/griffin/model/ExpressionNode.java [hunk 2]
```diff
@@ -167,15 +167,19 @@ public boolean equals(Object o) {
                 && Objects.equals(rhs, that.rhs);
     }
 
-    public boolean hasLeafs() {
-        return lhs != null && rhs != null;
+    public boolean noLeafs() {
```
Comment (marregui) [inline core/src/main/java/io/questdb/griffin/model/ExpressionNode.java:170]: why not call this method `x2`? 

not to have leafs == left leaf is null and right leaf is null.
in the condition only one null is required, so a better name would be 'maybeOneLeaf?'

File: core/src/test/java/io/questdb/test/griffin/SqlParserTest.java [hunk 3]
```diff
@@ -7583,7 +7621,7 @@ public void testSelectWildcardTabNoFrom() throws Exception {
                 "select * tab",
                 "create table tab (seq long)",
                 9,
-                "'from' expected"
+                "wildcard cannot have alias"
```
Comment (marregui) [inline core/src/test/java/io/questdb/test/griffin/SqlParserTest.java:7624]: Tested this and works fine:

```
select *, i, from t
select *, i from t
```

**Review by marregui — APPROVED** [review]

File: core/src/main/java/io/questdb/griffin/model/ExpressionNode.java (hunk 2 above)
Comment (bluestreak01) [reply core/src/main/java/io/questdb/griffin/model/ExpressionNode.java:170]: this was always reversed, e.g. used in `!hasLeafs()`, intellij kept complaining, so i reversed the original method in easiest way possible to stop the complaints

File: core/src/main/java/io/questdb/griffin/SqlOptimiser.java (hunk 1 above)
Comment (bluestreak01) [reply core/src/main/java/io/questdb/griffin/SqlOptimiser.java]: removed here https://github.com/questdb/questdb/pull/3684/commits/e62df8b72ce09273453bee6c5ebdeeb5f775f801

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 33 / 33 (100.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/SqlParser.java|27|27|100.00%|
|:large_blue_circle:|io/questdb/griffin/WhereClauseParser.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/model/ExpressionNode.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlOptimiser.java|3|3|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 99 / 101 (98.02%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/SqlParser.java|94|96|97.92%|
|:large_blue_circle:|io/questdb/griffin/WhereClauseParser.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/model/ExpressionNode.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlOptimiser.java|2|2|100.00%|

**Review by marregui — COMMENTED** [review]
```
create table from(from symbol, ts timestamp) timestamp(ts) partition by day wal;

select from, from from where from is not null;
```

BOOM!

**Comment by bluestreak01** [comment]
PostgreSQL requires quote marks around `from` when it is used in context of table name, e.g.

```sql
create table "from"("from" symbol, ts timestamp) timestamp(ts) partition by day wal;
select "from", from "from" where "from" is not null;
```

If quotes are omitted anywhere in those SQLs there will be syntax error (without position actually)

I feel this is a separate bug (unrelated to this change). We could fix this separately

**Comment by bluestreak01** [comment]
the issue is with enforcement that keywords should be quoted when they are used as column and table names. This is not part of this PR (pre-existing issue on master).

When SQL has correct syntax, e.g.

```sql
create table "from"("from" symbol, ts timestamp) timestamp(ts) partition by day wal;
select "from", from "from" where "from" is not null;
```

this PR does the job.

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 99 / 101 (98.02%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/SqlParser.java|94|96|97.92%|
|:large_blue_circle:|io/questdb/griffin/WhereClauseParser.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/model/ExpressionNode.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlOptimiser.java|2|2|100.00%|

**Review by ideoma — APPROVED** [review]

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 99 / 101 (98.02%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/SqlParser.java|94|96|97.92%|
|:large_blue_circle:|io/questdb/griffin/WhereClauseParser.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/model/ExpressionNode.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlOptimiser.java|2|2|100.00%|

**Pull request merged**